### ISSUE TITLE
Polish making watchlist not rely only on uuid

### ISSUE THREAD

**Issue by shlomi-nx** [nutanix/papiea#449]
@joshua-berry-ntnx  and I spent a long time identifying this bug. Please take on our progress in https://github.com/nutanix/papiea/pull/446.

See comments in https://github.com/nutanix/papiea/pull/446/commits/7ffdb1b7197c0ff5791da634fd6fcc3dfafd9793

Please go over the code and make sure everything is polished. Also add test cases such as:
1. two providers, each with a kind and an entity with the same id - update spec of both and ensure both are driven intentfully to completion
2. Same as (1) but both providers have a kind with the same name, and an entity with the same id
3. One provider with two kinds each has an entity with the same uuid. Make intentful change see that they are being resolved.

Add more test cases to make sure this is robust.

**Comment by Emulebest** [comment]
@shlomi-nx regarding 3. as of now, uuid is unique across the whole provider
```
{ "metadata.uuid": 1, "metadata.provider_prefix": 1 },
{ name: "provider_specific_entity_uuid", unique: true },
```
would you like to change that?

**Comment by shlomi-nx** [comment]
Oh right, I think so. @joshua-berry-ntnx thoughts?

**Comment by joshua-berry-ntnx** [comment]
Yes, the unit of uniqueness should be the tuple `(provider, kind, uuid)`. (And this should probably also be the same set of fields we use in a reference.)

I think this is slightly different from what we put in #446—IIRC we put version in there as well, and it probably shouldn't be (but maybe I'm misremembering?).

**Comment by shlomi-nx** [comment]
Regarding #446, we did put the version as well in the entity's identity: https://github.com/nutanix/papiea/blob/e3d684e782eae691e9096606c49ae3430c18f853/papiea-engine/src/intentful_engine/watchlist.ts#L11-L16
But I think that this is OK. In the worse case it will attempt to resolve an entity twice, the second time will be a noop if the first succeeded.

**Comment by Emulebest** [comment]
resolved by #446
